File: miband-sdk/src/main/java/com/khmelenko/lab/miband/model/LeParams.java
package com.khmelenko.lab.miband.model;

/**
 * LE parameters
 *
 * @author Dmytro Khmelenko
 */
public final class LeParams {

    public int mConnIntMin;
    public int mConnIntMax;
    public int mConnInt;

    public int mLatency;
    public int mTimeout;
    public int mAdvInt;

    /**
     * Creates an instance of LeParams from byte data
     *
     * @param data Byte data
     * @return Instance of the object
     */
    public static LeParams fromByte(byte[] data) {
        LeParams params = new LeParams();

        params.mConnIntMax = 0xffff & (0xff & data[0] | (0xff & data[1]) << 8);
        params.mConnIntMax = 0xffff & (0xff & data[2] | (0xff & data[3]) << 8);
        params.mLatency = 0xffff & (0xff & data[4] | (0xff & data[5]) << 8);
        params.mTimeout = 0xffff & (0xff & data[6] | (0xff & data[7]) << 8);
        params.mConnInt = 0xffff & (0xff & data[8] | (0xff & data[9]) << 8);
        params.mAdvInt = 0xffff & (0xff & data[10] | (0xff & data[11]) << 8);

        params.mConnIntMin *= 1.25;
        params.mConnIntMax *= 1.25;
        params.mAdvInt *= 0.625;
        params.mTimeout *= 10;

        return params;
    }
}

File: miband-sdk-kotlin/src/main/java/com/khmelenko/lab/miband/model/LeParams.kt
package com.khmelenko.lab.miband.model

/**
 * LE parameters
 *
 * @author Dmytro Khmelenko
 */
class LeParams {

    var connIntMin: Int = 0
    var connIntMax: Int = 0
    var connInt: Int = 0

    var latency: Int = 0
    var timeout: Int = 0
    var advInt: Int = 0

    companion object {
        fun fromByte(data: ByteArray): LeParams {
            val params = LeParams()
            params.connIntMax = 0xffff and (0xff and data[0].toInt() or (0xff and data[1].toInt() shl 8))
            params.connIntMax = 0xffff and (0xff and data[2].toInt() or (0xff and data[3].toInt() shl 8))
            params.latency = 0xffff and (0xff and data[4].toInt() or (0xff and data[5].toInt() shl 8))
            params.timeout = 0xffff and (0xff and data[6].toInt() or (0xff and data[7].toInt() shl 8))
            params.connInt = 0xffff and (0xff and data[8].toInt() or (0xff and data[9].toInt() shl 8))
            params.advInt = 0xffff and (0xff and data[10].toInt() or (0xff and data[11].toInt() shl 8))

            params.connIntMin = (params.connIntMin * 1.25).toInt()
            params.connIntMax = (params.connIntMax * 1.25).toInt()
            params.advInt = (params.advInt * 0.625).toInt()
            params.timeout *= 10

            return params;
        }
    }

}
